### PR TITLE
dev -> main. Schema changes; switch to flask-only; multi-label model

### LINKED ISSUE #20: Claim URLs should include timestamp
### Describe the bug

Currently, all URLs of claims found within a video are identical. Instead, each claim should link to the point in the video where the claim is made.

### To Reproduce

Steps to reproduce the behaviour:

1. Go to [raphael](https://raphael.fullfact.org/?q=8PqlzOVfciQ)
2. Hover over (or click on) any of the 5 claims
3. Note that each ends `&t=0`

### Expected behaviour

Each link URL should end with the timestamp associated with the claim.
(Acutally, we should probably link a few seconds earlier).

### LINKED ISSUE #21: When adding a new video, tool should accept full URL or just the video ID
### Requirements

Users should be able to copy/paste the complete URL of a YouTube video such as:
* https://www.youtube.com/watch?v=tQ6l56-ZkeA

Or just the video id such as:
* tQ6l56-ZkeA

Stretch goal (only if it's not much work) - allow more complex URLs, such as:
* https://youtu.be/tQ6l56-ZkeA?feature=shared&t=43
which is generated by the "share" option from within YouTube.

### LINKED ISSUE #23: Show video title in Previous Searches as well as video id
### Overview

Currently, the list of Previous Searches just shows the video ID and links to the list of extracted claims.

As a user demonstrating this tool, 
I want to see the title of each processed video,
so I can easily find the one I'm looking for.

It's possible that videos may have the same/similar titles, so including the video id would be helpful.

If the title is very long, it could be truncated. (Not sure if that ever happens though.)

### LINKED ISSUE #42: Show raw transcript beside paraphrased claims
### Overview

For clarity, the tool shows paraphrased version of extracted claims. The first thing any fact checker would want to do is to read the actual claim made in context. (This is partly to check the model hasn't hallucinated any extra meaning into the claim.) 

### Requirements

After a video has been processed and a list of claims is displayed, also display the entire transcript in a scrollable area to one side. 

When a claim is clicked, the video should start playing from that point (as per [issue 20](https://github.com/FullFact/health-misinfo-shared/issues/20)) AND also the transcript display should jump to that point in the transcript and highlight the raw text of the claim.

In this mockup, the user has clicked on a claim on the left and can see the raw transcript on the right with the same claim highlighted. (The video should be playing somewhere too.)
raphael_transcript_linking

### Notes and additional information

I am not a UI expert! It may be that starting a video AND jumping the text at the same time is confusing or obscure. An alternative might be to add a little "jump to transcript" button beside each extacted claim, maybe with a second button saying "jump to video".

### LINKED ISSUE #61: Indicate degree of checkworthiness
### Overview

Our prompt & model should identify claims worth checking from a video, but should also indicate _how_ checkworthy it is. [One form of the prompt](https://github.com/FullFact/health-misinfo-shared/blob/0f671fa9edf91e2bfb531fe2bcafdff4176350d9/src/health_misinfo_shared/prompts.py#L169) (for example) labels each claim as one of: 
"not worth checking", "worth checking", "may be worth checking"

These labels should be reflected in the front end by grouping the claims by that "summary" label so that the claims "worth checking" are shown at the top, clearly demarkated (a text label and/or maybe a background colour?); then the "may be worth checking" as a separate block below; and finally the "not worth checking" at the bottom. 

The on-screen labels should reflect our caveats; e.g. in order: "mostly likely to be worth checking", "may be worth checking" and "less likely to be worth checking" for these three labels.

### Requirements

- [ ] make this prompt the default so that newly-processed videos also get the "summary" label
- [ ] group claims by this label and sort the groups from most to least checkworthy down the screen
- [ ] choose and add suitable text labels for each group

### Notes and additional information

Later on, we might have more than 3 groups, poss. even a continuous score. But let's see how this crude 'traffic light' sorting works first.

### LINKED ISSUE #62: Handle multiple analyses of the same video
### Describe the bug

Currently, if someone enters the video id and clicks analyse for a video that has already been analysed, the model will extract the same (or similar) claims again and add them to the list. 

Instead, we should list the video twice (or more times) on the main page, each linking to its associated set of claims.

This is especially important as we update the prompts/models etc. and may want to compare the same video before and after a change.

### To Reproduce

Steps to reproduce the behaviour:

1. Go to raphael
2. enter a video id and click analyse
3. review the extracted claims
4. enter THE SAME video id and click analyse
5. observe more claims have appeared in the list, roughly doubling it in size

### Expected behaviour

If the same id occurs twice in the list of analysed videos, distinguish by a version number. E.g. 

* 4WAFHXdTMbY
* 4WAFHXdTMbY (v2)
* 4WAFHXdTMbY (v3)

etc.

### Additional context

For reference: in Live, if the same YouTube video is analysed twice, two versions are shown. This is what we want here.

### LINKED ISSUE #66: process_video should return an iterator
`process_video` runs chunks of text through an LLM, and appends the results to a list which it eventually returns. For long videos, it can take a long time to complete. If there’s an error along the way, this can mean none of the results get saved to the database.

Instead of appending to a big list and returning that, it would be preferable to yield results as it gets them.

### LINKED ISSUE #70: Change main inference model to multi-label version
### Overview

We have developed a prompt & model that generates 5 labels for each claim:
understandability, type_of_claim , type_of_medical_claim , support , harm

We should switch the model used in the MVP to use this version instead, as a precursor to fixing #61 as it will allow us to display and sort by the degree of checkworthiness.

We potentially have the option of using a fine-tuned Gemini model here, or using in-context learning (i.e. putting the training data in the prompt).  The results should be broadly similar so let's deploy whatever is simplest to deploy!

### Requirements

- [ ] switch the model in `vertex.py/generate_reponse()` to the new model
- [ ] make sure the output (a JSON object) is stored in the `inferred_claims` table - the output will need to be converted from JSON to a single string first

### LINKED ISSUE #71: Database schema revisions
The current database schema (https://github.com/FullFact/health-misinfo-shared/issues/62#issuecomment-2118785064) has a few gaps. We’d like to store the following things:
 * multiple runs of the same youtube video (e.g. using different prompts) (#62)
 * the raw text that a claim was extracted from (#42)
 * offset end timestamps for inferred claims
 * training claims for videos that may not have been run through claim extraction

Here’s a proposed revised schema. This is WIP, and it may also be more complicated than we really need. But hopefully it should capture the things mentioned above.

```mermaid
erDiagram
  youtube_videos ||--o{ claim_extraction_runs : runs
  youtube_videos ||--o{ training_claims : claims
  youtube_videos {
    text id
    text metadata
    text transcript
  }

  claim_extraction_runs ||--o{ inferred_claims : claims
  claim_extraction_runs {
    integer id PK
    text youtube_id FK
    text model
    text status
  }

  inferred_claims {
    integer id PK
    integer video_id FK
    text claim
    text raw_transcript_text
    text label
    integer offset_start_s
    integer offset_end_s
  }

  training_claims {
    integer id PK
    text youtube_id FK
    text claim
    text label
  }
```

### LINKED ISSUE #77: Replace react with flask-only version

### PR DESCRIPTION
**Pull request by andylolz**
This merges `dev` into `main`, which includes lots of fixes:

Fixes #20.
Fixes #21.
Fixes #23.
Refs #42 (makes progress on this one, but possibly we should keep it open because this doesn’t exactly address it).
Fixes #62.
Fixes #66.
Fixes #71.
Fixes #77.

With #76 included in here, this also:

Fixes #61.
Fixes #70.

-----------------

## How to deploy

1. merge this branch
2. checkout main and pull
3. the nginx_docker ansible playbook will need to be run, because this PR includes nginx changes. [Instructions in the README](https://github.com/FullFact/health-misinfo-shared/tree/dev?tab=readme-ov-file#deployupdate-a-server)
4. delete the database
5. run [the deployment workflow](https://github.com/FullFact/health-misinfo-shared/actions/workflows/ci.yml) on the main branch

I **think** that should do the trick? 

-----------------

## Pull request checklist

- [x] I have linked my PR to an issue
- [x] I’ve used [conventional commits](https://github.com/FullFact/automation-docs/wiki/Conventional-commits)
- [x] My branch is up-to-date with `main`
- [ ] Where appropriate, I have added or updated tests
- [x] Where appropriate, I have [updated documentation](https://github.com/FullFact/automation-docs/) to reflect my changes

### CODE REVIEW

File: src/health_misinfo_shared/vertex.py [hunk 1]
```diff
@@ -61,14 +61,13 @@ def process_video(video_id: str, transcript: list[dict]) -> list[dict]:
             for found_claim in llm_response:
                 found_claim["video_id"] = video_id
                 found_claim["chunk"] = chunk["text"]
-                found_claim["offset_s"] = chunk["start_offset"]
-                found_claim["offset_end_s"] = chunk["end_offset"]
+                found_claim["offset_start_s"] = float(chunk["start_offset"])
+                found_claim["offset_end_s"] = float(chunk["end_offset"])
                 print(found_claim)
```
[the remark this answers is not in the document]
Comment (andylolz) [reply src/health_misinfo_shared/vertex.py:66]: That’s one for @dcorney.

File: src/raphael_backend_flask/db.py [hunk 2]
```diff
@@ -1,58 +1,138 @@
...
+        ),
+    )
+    return result[0]["id"]
+
+
+def update_claim_extraction_run(
+    video_id: str,
+    **kwargs: str,
+) -> None:
+    update_cols = ", ".join([f"{k} = ?" for k in kwargs.keys()])
```
[the remark this answers is not in the document]
Comment (andylolz) [reply src/raphael_backend_flask/db.py:73]: Strong agree – it’s really horrible

File: src/raphael_backend_flask/db.py [hunk 3]
```diff
@@ -1,58 +1,138 @@
...
+
+
+table_youtube_videos = """
+    CREATE TABLE youtube_videos (
         id TEXT PRIMARY KEY,
-        url TEXT,
         metadata TEXT,
-        transcript TEXT,
-        status TEXT
+        transcript TEXT
```
[the remark this answers is not in the document]
Comment (andylolz) [reply src/raphael_backend_flask/db.py:88]: That’s a very good shout

[the remark this answers is not in the document]
Comment (andylolz) [reply src/raphael_backend_flask/db.py:88]: I’ll add some inline comments here, I guess

File: src/raphael_backend_flask/db.py [hunk 4]
```diff
@@ -1,58 +1,138 @@
...
     """
 
 table_inferred_claims = """
     CREATE TABLE inferred_claims (
         id INTEGER PRIMARY KEY AUTOINCREMENT,
-        video_id TEXT,
+        run_id INTEGER,
         claim TEXT,
+        raw_sentence_text TEXT,
         label TEXT,
```
Comment (andylolz) [inline src/raphael_backend_flask/db.py:109]: I wonder if this should be renamed `labels` now, given it now contains multiple labels in multiple categories… WDYT, @ff-dh?

Comment (ff-dh) [reply src/raphael_backend_flask/db.py:109]: +1 fair shout

File: src/raphael_backend_flask/db.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (andylolz) [reply src/raphael_backend_flask/db.py:89]: Gah – hadn’t occurred to me until I tested it, but inlining comments causes this error:
```
sqlite3.OperationalError: unrecognized token: "#"
```

File: src/raphael_backend_flask/db.py (hunk 4 above)
Comment (andylolz) [reply src/raphael_backend_flask/db.py]: Okay, should be sorted

File: src/health_misinfo_shared/vertex.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (dcorney) [reply src/health_misinfo_shared/vertex.py:66]: Yes, it's a rogue debug statement. Good to remove.

[the remark this answers is not in the document]
Comment (andylolz) [reply src/health_misinfo_shared/vertex.py]: Great – thanks. Sorted

**Comment by ff-dh** [comment]
I think this is ready to approve?

**Review by ff-dh — APPROVED** [review]

**Comment by dcorney** [comment]
I'm also happy for this to be merged! I haven't had time to do a line-by-line review, but the dev branch has been working locally, and I can't see anything here that would break it. Let us merge!

**Review by dcorney — APPROVED** [review]

**Pull request merged**